File: leetcode/course-schedule-iv.py
# 1462. Course Schedule IV
# 🟠 Medium
#
# https://leetcode.com/problems/course-schedule-iv/
#
# Tags: Depth-First Search - Breadth-First Search - Graph - Topological Sort

import timeit
from typing import List


# Use topological sort to create an array where each entry i is a set
# with all the dependencies of the i-th course, direct and indirect ones,
# once we have that array, we can perform each query in O(1), given the
# constraints of the problem, it is better to use a bit more time
# building this data structure, even if the complexity of that is high,
# O(n^3+p) because we know that n <= 100 is a small value. After that
# the queries, of which we could have up to 10^4, can run in O(1).
#
# Time complexity: O(n^3 + p) - The last loop where we go along the
# topographical order processing nodes and creating the all_dependencies
# array has 3 nested loops all of which have an upper limit of n. The
# third loop is implicit, the copy of set elements from a dependency to
# the current node.
# Space complexity: O(n^2) - The arrays of size n have elements which
# in turn are sets of size n.
#
# Runtime 738 ms Beats 87.53%
# Memory 18.1 MB Beats 11.16%
class TopologicalSort:
    def checkIfPrerequisite(
        self,
        numCourses: int,
        prerequisites: List[List[int]],
        queries: List[List[int]],
    ) -> List[bool]:
        # Create two hashmaps with indegree and outdegree. O(n)
        direct_dependencies = [set() for _ in range(numCourses)]
        is_depended_on = [set() for _ in range(numCourses)]
        # Iterate over the prerequisites adding them to the hashmaps,
        # this will take O(p)
        for depended, dependent in prerequisites:
            direct_dependencies[dependent].add(depended)
            is_depended_on[depended].add(dependent)
        # Copy the dependencies, direct dependencies only has p
        # dependencies, this is also O(p)
        all_dependencies = [s.copy() for s in direct_dependencies]
        # Process courses that don't have any remaining dependencies. O(n)
        stack = [i for i in range(numCourses) if not direct_dependencies[i]]
        # Iterate over the courses on the stack, courses go on the stack
        # once we have processed oll their dependencies, this loop will
        # run a maximum of n times O(n)
        while stack:
            current = stack.pop()
            # Iterate over all courses that depend on this one at this
            # point, is_depended_on includes only direct dependencies
            # even though it is likely to be much smaller than n, that
            # is its upper bound. O(n)
            for dependent in is_depended_on[current]:
                # Union the indirect dependencies from this dependency.
                # A node could be dependent in all others, O(n)
                all_dependencies[dependent] |= all_dependencies[current]
                direct_dependencies[dependent].remove(current)
                # Once we have processed all dependencies, we can
                # process this course.
                if not direct_dependencies[dependent]:
                    stack.append(dependent)
        return [u in all_dependencies[v] for u, v in queries]


def test():
    executors = [TopologicalSort]
    tests = [
        [2, [], [[1, 0], [0, 1]], [False, False]],
        [2, [[1, 0]], [[0, 1], [1, 0]], [False, True]],
        [3, [[1, 2], [1, 0], [2, 0]], [[1, 0], [1, 2]], [True, True]],
        [
            5,
            [[3, 4], [2, 3], [1, 2], [0, 1]],
            [[0, 4], [4, 0], [1, 3], [3, 0]],
            [True, False, True, False],
        ],
    ]
    for executor in executors:
        start = timeit.default_timer()
        for _ in range(1):
            for col, t in enumerate(tests):
                sol = executor()
                result = sol.checkIfPrerequisite(t[0], t[1], t[2])
                exp = t[3]
                assert result == exp, (
                    f"\033[93m» {result} <> {exp}\033[91m for"
                    + f" test {col} using \033[1m{executor.__name__}"
                )
        stop = timeit.default_timer()
        used = str(round(stop - start, 5))
        cols = "{0:20}{1:10}{2:10}"
        res = cols.format(executor.__name__, used, "seconds")
        print(f"\033[92m» {res}\033[0m")


test()

File: leetcode/course-schedule-iv.rs
// 1462. Course Schedule IV
// 🟠 Medium
//
// https://leetcode.com/problems/course-schedule-iv/
//
// Tags: Depth-First Search - Breadth-First Search - Graph - Topological Sort

struct Solution;
impl Solution {
    // We will use a hashmap of hash sets to store all the dependencies of each course,
    // both direct and indirect, first create an adjacency list that we will use to do
    // a dfs for each course starting by their direct dependencies and following the
    // dependency chain, using the adjacency list, to collect all the dependencies.
    // We will use the same hashmap as a memo object to avoid recomputing one course's
    // dependencies multiple times, once the course is present in the set, we know that
    // its dependencies have been computed. To create the result vector, we iterate over
    // the queries and simply check if the query first node is a dependency of the second
    // node, which can be done in O(1) using the hashmap created in the previous step.
    //
    // Time complexity: O(n^3 + p) - Where p is the number of prerequisites and n is the
    // number of courses, first we iterate over the prerequisites to create an adjacency
    // list, then we call dfs for all courses, dfs will visit all the courses neighbors
    // and, for each, it may need to copy n dependencies to its dependency hash set.
    // Space complexity: O(n^2) - The dependencies hashmap will have n entries, each entry
    // is a hash set that could have up to n entries.
    //
    // Runtime 65 ms Beats 33.33%
    // Memory 3 MB Beats 100%
    pub fn check_if_prerequisite(
        num_courses: i32,
        prerequisites: Vec<Vec<i32>>,
        queries: Vec<Vec<i32>>,
    ) -> Vec<bool> {
        use std::collections::{HashMap, HashSet};
        let n = num_courses as usize;
        // Create an adjacency list.
        let mut adj = vec![vec![]; n];
        for pre in prerequisites.iter() {
            adj[pre[1] as usize].push(pre[0] as usize);
        }
        let mut dep = HashMap::<usize, HashSet<usize>>::new();
        // Define a DFS function that finds and returns all dependencies for a given course.
        fn dfs(adj: &Vec<Vec<usize>>, c: usize, dep: &mut HashMap<usize, HashSet<usize>>) {
            // Use dep as a memo, if the value has been computed, return.
            if dep.contains_key(&c) {
                return;
            }
            let mut deps = HashSet::new();
            for nei in adj[c as usize].iter() {
                // This neighbor is a dependency.
                deps.insert(*nei);
                // Collect the dependencies if not in the memo.
                dfs(&adj, *nei, dep);
                for val in dep.entry(*nei).or_default().iter() {
                    deps.insert(*val);
                }
            }
            dep.insert(c, deps);
        }
        // Collect the list of dependencies for each course.
        for i in 0..n {
            dfs(&adj, i, &mut dep);
        }
        (0..queries.len())
            .map(|i| {
                dep.entry(queries[i][1] as usize)
                    .or_default()
                    .contains(&(queries[i][0] as usize))
            })
            .collect()
    }
}

// Tests.
fn main() {
    assert_eq!(
        Solution::check_if_prerequisite(2, vec![], vec![vec![1, 0], vec![0, 1]]),
        vec![false, false]
    );
    assert_eq!(
        Solution::check_if_prerequisite(2, vec![vec![1, 0]], vec![vec![0, 1], vec![1, 0]]),
        vec![false, true]
    );
    assert_eq!(
        Solution::check_if_prerequisite(
            3,
            vec![vec![1, 2], vec![1, 0], vec![2, 0]],
            vec![vec![1, 0], vec![1, 2]]
        ),
        vec![true, true]
    );
    assert_eq!(
        Solution::check_if_prerequisite(
            5,
            vec![vec![3, 4], vec![2, 3], vec![1, 2], vec![0, 1]],
            vec![vec![0, 4], vec![4, 0], vec![1, 3], vec![3, 0]]
        ),
        vec![true, false, true, false]
    );
    println!("All tests passed!")
}
